src/core/experimental/refcount.d

```diff
@@ -0,0 +1,782 @@
+        }
+        else
+        {
+            mixin("return cast(CounterType)(*(cast(CounterType*) rc)" ~ op ~ "val);");
+        }
+    }
+
+    /**
+     * Creates a new `__RefCount` instance. It's memory is internally managed with
+     * malloc/free.
+     *
+     * Params:
+     *      _ = an unused int value; required because structs don't have a
+     *          user defined default constructor
+     */
+    @nogc nothrow pure @trusted scope
+    this(this Q)(int)
+    {
+        /* We allocate a `size_t` chunk that will save as our support. We
+         * logically split the chunk into two `uint`s, using only one of the
+         * two as our counter, depending if we are creating an immutable `__RefCount`
+         * or not. The logic is as follows:
+         *  - if we are creating an immutable RC, then a pointer to the first
+         *    `uint` (aligned at 8) will serve as the reference count. On this
+         *     counter we will only perform atomic operations, as immutable can
+         *     be used in shared contextes.
+         *  - if we are creating a const/mutable RC, then a pointer to the second
+         *    `uint` (aligned at 4) will serve as the reference count.
+         */
+        CounterType* support = cast(CounterType*) pureAllocate(2 * CounterType.sizeof);
+        static if (is(Q == immutable))
+        {
+            *support = 0;
+            rc = cast(immutable CounterType*) support;
+        }
+        else
+        {
+            *(support + 1) = 0;
+            rc = cast(CounterType*) (support + 1);
+        }
+        addRef();
+    }
+
+    private enum copyCtorIncRef = q{
+        rc = rhs.rc;
+        assert(rc == rhs.rc);
+        if (rhs.isInitialized())
+        {
+            assert(isShared() == rhs.isShared());
+            addRef();
+        }
+    };
+
+    /**
+     * Copy constructs a mutable `__RefCount` from a mutable reference, `rhs`.
+     * This increases the reference count.
+     */
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs)
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    // { Get a const obj
+
+    /**
+     * Copy constructs a const `__RefCount` from a mutable reference, `rhs`.
+     * This increases the reference count.
+     */
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+     * Copy constructs a const `__RefCount` from a const reference, `rhs`.
+     * This increases the reference count.
+     */
+    @nogc nothrow pure @safe scope
+    this(return scope const ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+     * Copy constructs a const `__RefCount` from an immutable reference, `rhs`.
+     * This increases the reference count.
+     */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) const
```

Comment:
Is there any reason all of these constructors aren't just a template?

Reply:
Can't use templates with the copy constructor.

Reply:
Isn't this a major design deficient in copy constructor? CC @RazvanN7

Reply:
> I still don't see how this fits into the larger picture, probably because this PR doesn't explain what the long-term reference counting plan is. Do users have to add this `__RefCount` field and supporting boiler-plate to their types any time they want to have reference-counted semantics? Or is there some future compiler integration planned where the boilerplate is added by the compiler? How is this different from Phobos's `RefCounted` type? Can a `__RefCount` field also be added to classes?

This is the basic mechanism for allowing reference counted objects usable with all attributes and qualifiers. Libraries can use it as a building block. There are no current plans for automatic integration in generated code, though that's a future possibility.